### PR TITLE
glTF schema v4 upgrade and inheritance issue fix

### PR DESCRIPTION
**Pull request by icechen1**
There's an issue in the gltf schema related to inheritance that is causing schema validators to fail (I verified this with [rapidjson](https://github.com/miloyip/rapidjson/) and [json-schema-validator](https://github.com/daveclayton/json-schema-validator). A description of the problem can be found [here](http://stackoverflow.com/questions/27410216/json-schema-and-inheritance) and [here](http://stackoverflow.com/questions/22689900/json-schema-allof-with-additionalproperties).

I implemented a workaround that fixes the issue (hopefully a future version of json schema will provide a more elegant solution). The solution consists of passing the name of the fields defined in the "parent" down to the children and define them as empty objects. The advantage of this approach is that there is no duplication involved and the actual schema of the fields in question is only defined in the parent. This way the schema will not fail at additionalProperties and validators seem to behave as expected.

I've also added v4 schema upgrades as well. See https://github.com/json-schema/json-schema/wiki/ChangeLog for changes between schema v3 and v4

### CODE REVIEW

**Comment by pjcozzi** [comment]
Thanks for the workaround, @icechen1!

@lexaknyazev this is OK with me, does it impact the 1.1 schema?

@icechen1 this is likely to impact the [C# loader generated from the schema](https://github.com/KhronosGroup/glTF/tree/master/loaders/CSharp) and [wetzel](https://github.com/AnalyticalGraphicsInc/wetzel) used to generate parts of the spec.  Did you test them?

We could also retarget this to a new `schema-improvements` branch (since I imagine JSON Schema 4 updates are coming? #252) and then evaluate changes to the C# loader and wetzel after all the schema updates.

**Comment by nehmebilal** [comment]
@pjcozzi We could certainly combine this PR and the V4 changes (which @icechen1 has already implemented) in one PR if you prefer? 

That being said, the changes proposed in this PR are fixes to actual issues in the schema and I feel that master could benefit from such fixes? What do you think?

**Comment by pjcozzi** [comment]
If the v4 changes are imminent, can we just wait so we only have to propagate changes to the 1.1 spec and doc tool?

Otherwise, if this is holding you up, we can merge and just wait on those other changes.

**Comment by nehmebilal** [comment]
Yes, v4 changes are ready. Should we add these changes to this PR? A different branch?

We can certainly wait but we're hoping to merge the changes before other changes to the GLTF schema are made to avoid diverging and having to deal with painful merges.

**Comment by icechen1** [comment]
I've tested the changes with the C# loader and Wetzel and noticed that the changes are breaking. C# causes crashes (due to key duplication) and Wetzel generates [blanks fields](https://cloud.githubusercontent.com/assets/3333944/20155451/cdb140a4-a680-11e6-91de-64f5114410aa.png) due to the workaround. 

I believe the workaround conforms the json schema [specs](http://json-schema.org/latest/json-schema-validation.html#rfc.section.5.22), so changes are needed on the loader and Wetzel in order to support this change. The C# loader seems to merge the schema referenced in 'extends' creating the duplicate field issue. However, that seems not to be the expected behavior according to the spec, rather, the 2 documents needs to be parsed independently.

**Comment by pjcozzi** [comment]
Adding the v4 changes to this branch would be OK.

**Comment by icechen1** [comment]
Okay, I've pulled in the v4 schema updates as well. Will update the PR name and description.

**Comment by sbtron** [comment]
@MattMcMullan could you take a look at the issues with C# loader that @icechen1 mentions above

**Comment by pjcozzi** [comment]
Looks good, thanks so much!  Also, thanks for updating the extensions.

Fixes #252.

Should we merge after this:

> @MattMcMullan could you take a look at the issues with C# loader that @icechen1 mentions above

The wetzel updates can be separate, of course.

File: specification/schema/accessor.schema.json [hunk 1]
```diff
@@ -1,12 +1,15 @@
...
+    "$schema" : "http://json-schema.org/draft-04/schema",
     "title" : "accessor",
     "type" : "object",
     "description" : "A typed view into a bufferView.  A bufferView contains raw binary data.  An accessor provides a typed view into a bufferView or a subset of a bufferView similar to how WebGL's `vertexAttribPointer()` defines an attribute in a buffer.",
-    "extends" : { "$ref" : "glTFChildOfRootProperty.schema.json" },
+    "allOf" : [ { "$ref" : "glTFChildOfRootProperty.schema.json" } ],
     "properties" : {
+        "name" : {},
+        "extensions" : {},
+        "extras" : {},
```
Comment (lexaknyazev) [inline specification/schema/accessor.schema.json]: Why do we have both `"allOf"` and `"name", "extensions", "extras"` here?

**Comment by lexaknyazev** [comment]
Some schemas have `"dependencies"`. Its format was [changed](https://github.com/json-schema/json-schema/wiki/ChangeLog#dependencies) with v4. 

@icechen1 Could you fix that also?

**Comment by MattMcMullan** [comment]
Sure, I'll take a look

File: specification/schema/accessor.schema.json (hunk 1 above)
Comment (nehmebilal) [inline specification/schema/accessor.schema.json]: The author of the json schema explains why in this [stackoverflow post](http://stackoverflow.com/questions/22689900/json-schema-allof-with-additionalproperties).

**Comment by icechen1** [comment]
@lexaknyazev Updated the schema with the updated `dependencies` and `required` syntax. Thanks for pointing it out!

**Comment by lexaknyazev** [comment]
@icechen1 Great!
Besides C# loader, is it ready for merge?

**Comment by icechen1** [comment]
`wetzel` also needs to be updated, but otherwise all the schema upgrades are pushed.

**Comment by lexaknyazev** [comment]
@icechen1 
Should we change `required` in [WEB3D_quantized_attributes.accessor.schema.json](https://github.com/icechen1/glTF/blob/fa22eb33f61731dd3f1997981dbf791ed3e857e4/extensions/Vendor/WEB3D_quantized_attributes/schema/WEB3D_quantized_attributes.accessor.schema.json)?

---

> `wetzel` also needs to be updated

CC @pjcozzi

**Comment by lexaknyazev** [comment]
Also [CESIUM_RTC.schema.json](https://github.com/icechen1/glTF/blob/fa22eb33f61731dd3f1997981dbf791ed3e857e4/extensions/Vendor/CESIUM_RTC/CESIUM_RTC.schema.json).

**Comment by pjcozzi** [comment]
> wetzel also needs to be updated, but otherwise all the schema upgrades are pushed

This can be merged first.

**Comment by icechen1** [comment]
@lexaknyazev Whoops, missed those two. I'll make the change now.

**Pull request merged**